### DISCUSSION TITLE
How to get ref with new gatsby-plugin-image?

### DISCUSSION THREAD

**Discussion by cedricdelpoux** [gatsbyjs/gatsby#30120] [Help]
With `gatsby-image` I could do:

```
({photo}) => {
  const photoRef = useRef()

  useEffect(() => {
   const image = photoRef.current.imageRef.current
   ....
  }, [photoRef.current])

  return (
        <GatsbyImage ref={photoRef} image={getImage(photo)} alt={alt} />
  )
}
```
But it does not work anymore. How to use refs with new `gatsby-plugin-image` plugin?

Thank you

**Comment by ascorbic** [comment]
Hi. There isn't currently a way to do this with the new component. What are you trying to do with it?

**Comment by cedricdelpoux** [comment]
I try to use [medium-zoom](https://github.com/francoischalifour/medium-zoom) library. 

I know there are some alternative with react like API like [react-medium-image-zoom](https://github.com/rpearce/image-zoom) for example but it's does not work neither. 

I would like to use the first one because this library is already installed with [gatsby-remark-images-medium-zoom](https://github.com/JaeYeopHan/gatsby-remark-images-medium-zoom) as a dependency.

It's too bad there is no way to do this with the new plugin because there are a lot a use case when we want to access images.

**Comment by joyfulelement** [comment]
For me, I also wanted to apply the medium zoom effect, as well as **creating a parallax effect for the responsive images** render from `<GatsbyImage>` component. 

One idea is to use React `ref` is to access `<img>` or `<picture>` elements within the `data-gatsby-image-wrapper` created by `<GatsbyImage>` component, e.g. the resulting HTML is:

```html
<div data-gatsby-image-wrapper="" class="gatsby-image-wrapper">
  <div aria-hidden="true" style="padding-top: 56.25%"></div>
  <div
    aria-hidden="true"
    data-placeholder-image=""
    style="opacity: 0; transition: opacity 500ms linear 0s; object-fit: cover"
  ></div>
  <picture>
    <source type="image/webp" srcset="..." sizes="100vw" />
    <img width="1" height="0.5625" data-main-image="" sizes="100vw"
    decoding="async" src="..." srcset=" " alt=""
  </picture>
</div>
```
Today, if we try access it with `ref`,  e.g.

```jsx
const test = () => {
    const heroImage = getImage(gatsbyImageData);
    const pictureRef = useRef(null);
    return (
        <GatsbyImage image={heroImage} alt={title} ref={pictureRef} />
    )
}
```

will simply get the following error.
```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
```
I understand the error. 

Not sure if anyone else has successfully created the parallax effect for the responsive images render from `<GatsbyImage>` component without needing `ref`? Any guidance would be much appreciated!

Perhaps we can submit a feature request for this? 
Thanks

**Comment by stephzero1** [comment]
I think, you could try to wrap the GatsbyImage component into a div and then apply the ref to the div:
`
<div ref={ref}>
    <GatsbyImage......
</div>
`

**Comment by rpearce** [comment]
@cedricdelpoux I'm late to this conversation, but with my recent release of `react-medium-image-zoom@5` _and_ this not working because of the reasons listed in https://github.com/rpearce/react-medium-image-zoom/issues/347, I am determined to figure it out somehow!

My `gatsby develop` solution works, but it doesn't work for `gatsby build` output because the `StaticImage` `onLoad` callback is never called.

I'm working on it!

**Comment by rpearce** [comment]
I believe I've fixed this for `gatsby-plugin-image` in `react-medium-image-zoom@5.0.2-beta.0` and expect to release it in `5.0.2` soon. Again, here's the issue for tracking purposes: https://github.com/rpearce/react-medium-image-zoom/issues/347

**Reply by rpearce** [reply]
Update: `react-medium-image-zoom@5.0.2` is now released.
